[WJ-506] Implement generic wikitext backend interface

emmiegit opened the pull request:
I've created a research document describing the uses of `WikiTransformation` here: https://scuttle.atlassian.net/wiki/spaces/WD/pages/1210712065/WikiTransformation+Usages

This PR transforms those usages into calling through the `WikitextBackend` abstract class, which then delegates the actual work to one of three backends:
* `TextWikiBackend` -- sends calls to `Text_Wiki`, the original Wikidot parser/render
* `FtmlBackend` -- sends calls to libftml, our eventual replacement
* `DummyBackend` -- trivial implementation for testing

You can change which backend is used by setting `wikitext_backend` in `wikijump.ini` to one of `text_wiki`, `ftml`, or `dummy`.

This also has a lot of miscellaneous changes to refactor Wikidot code and fix issues with local deployment:
* [WJ-516](https://scuttle.atlassian.net/browse/WJ-516), which removes the social bookmarks integration
* The "smileys" parse rule, which isn't used and we don't have the assets for anyways
* The page preview module, since it has some really annoying assumptions for `WikiTransformation` that would've made `WikitextBackend` much more annoying, and we're replacing it with our new editor Sheaf anyways.

Screenshot showing the `FtmlBackend` in use as the wikitext parser/renderer. (Headers are plaintext since they aren't added yet). I've tested all three backends and they work as expected.
![backend use](https://user-images.githubusercontent.com/8848022/118071778-6d86dd80-b376-11eb-8a66-589f6bd84618.png)

This implementation of `WikitextBackend` does _not_ convey parser warnings, CSS, or backlinks. These will be addressed in future PRs.

emmiegit committed:
Remove unused field.

emmiegit committed:
Create WikitextBackend interface and stubs.

emmiegit committed:
Add function version() to WikitextBackend.

emmiegit committed:
Construct WikiTransformation.

emmiegit committed:
Remove commented code in WikiTransformation.

emmiegit committed:
Remove "antiwiki" code.

emmiegit committed:
Apply some recommended PhpStorm refactors.

emmiegit committed:
Properly handle null from selectByPrimaryKey().

emmiegit committed:
Add missing declaration to GlobalProperties.

danieltharp reviewed (approved):
Sure, let's give it a shot.

emmiegit committed:
Don't use GlobalPropertiesException.

emmiegit committed:
Remove unused function setTransformationFormat().

emmiegit committed:
Initialize $wiki in constructor, remove resetWiki().

emmiegit committed:
Note on apparently-unused method.

emmiegit committed:
Move purifyHtml() to new utility class.

emmiegit committed:
Fix WikitextBackend errors.

emmiegit committed:
Remove 'Social' rule from Text_Wiki.

emmiegit committed:
Remove 'Smiley' rule from Text_Wiki.

emmiegit committed:
Start to add page info to WikitextBackend.

emmiegit committed:
Move ftml FFI classes to a subdirectory.

Avoid causing confusion with general wikitext service classes.

emmiegit committed:
Add PHP PageInfo storage class.

emmiegit committed:
Add getCategory() method to PageInfo.

emmiegit committed:
Add getPageSlug() method to PageInfo.

emmiegit committed:
Add getFullPageSlug() method to PageInfo.

emmiegit committed:
Set WikiTransformation page if it has PageInfo.

emmiegit committed:
Add ParseRenderMode enum.

emmiegit committed:
Refactor out setPageUnixName().

Remove some associated duplicated code.

emmiegit committed:
Create non-FFI output classes.

This splits out the previously FFI-based data classes, and instead
moves FFI conversion logic over to FFI/OutputConversion.php

emmiegit committed:
Create HtmlMetaType enum and use that.

emmiegit committed:
Use [] instead of the deprecated array().

emmiegit committed:
Add stubbed render methods.

emmiegit committed:
Add private constructor and phpdoc for FtmlFfi::_init().

emmiegit committed:
Implement FtmlBackend render methods.

emmiegit committed:
Conform to PSR-12.

emmiegit committed:
Refactor HtmlUtilities::purify().

Horribly overwrought piece of Wikidot jank.

emmiegit committed:
Add ParseRenderMode::TABLE_OF_CONTENTS.

emmiegit committed:
Replace WikiTransformation usages with getWikitext().

emmiegit committed:
Remove unused $pageUnixName field.

emmiegit committed:
Remove useless comment.

emmiegit committed:
Rename getWikitext() -> getWikitextBackend().

emmiegit committed:
Rename DummyBackend and give it some basic output.

emmiegit committed:
Add comment explaining $pageInfo requirements.

emmiegit committed:
Remove unneeded use statements.

And some commented-out code.

emmiegit committed:
Remove PagePreviewModule.

Adds a number of complications which aren't worth solving
if we're going to replace the entire editing experience.

emmiegit committed:
Remove unused processSource().

emmiegit committed:
Return directly.

emmiegit committed:
Split out template assembly into its own file.

emmiegit committed:
Add comment on the unused setMode() method.

emmiegit committed:
Remove stray newlines.

emmiegit committed:
Fix throws declaration.

emmiegit committed:
Add declare(strict_types = 1).

emmiegit committed:
Rename helper ListPages method.

emmiegit committed:
Use closures instead of dynamic dispatch.

This way PhpStorm knows those methods are being used.

emmiegit committed:
Replace WikiTransformation usage with LegacyTemplateAssembler.

emmiegit committed:
Collapse method call.

emmiegit committed:
Collapse setPageSlug() method.

emmiegit committed:
Remove awiki mode.

emmiegit committed:
Revert "Remove unused processSource()."

This reverts commit 8922f4084c926f463d4babc4babccccf45cf2b5f.

I forgot that TextWikiBackend->renderHtml is stubbed.

emmiegit committed:
Implement TextWikiBackend->renderHtml().

emmiegit committed:
Set parse mode in TextWikiBackend constructor.

emmiegit committed:
Remove extraneous newline.

emmiegit committed:
Remove $transformationFormat.

It's always 'xhtml', no point in having this.

emmiegit committed:
Minor Text_Wiki/Wiki refactoring.

emmiegit committed:
Remove commented-out parse rules.

emmiegit committed:
Remove renderConf indirection, always assume xhtml.

emmiegit committed:
Remove formatConf indirection, always assume xhtml.

emmiegit committed:
Remove $currentFormat field.

emmiegit committed:
Fix syntax error.

emmiegit committed:
Remove unused insertRule() method.

emmiegit committed:
Remove unused rule methods.

emmiegit committed:
Miscellaneous fixes.

emmiegit committed:
Fix feature flag error.

emmiegit committed:
Record invalid mode in exception message.

emmiegit committed:
Social rule is already disabled.

emmiegit committed:
Remove social bookmarks code.

emmiegit committed:
Remove ancient social bookmark assets.

emmiegit committed:
Add DEADCODE entry for WJ-516.

emmiegit committed:
Add aliases for en wikipedia.

emmiegit committed:
Fix locale for postgres builds.

emmiegit committed:
Fix xdebug mode.

emmiegit committed:
Add mutable bind mount for wikijump local storage.

emmiegit committed:
Don't override child type in TrustProxies.php

emmiegit committed:
Handle potentially undefined $thread.

emmiegit committed:
Use helper method for converting $page to PageInfo.

emmiegit committed:
Instantiate PageInfo objects for rendering.

emmiegit committed:
Remove extra argument.

emmiegit committed:
Prevent creation of static utility class.

emmiegit committed:
Add dummy PageInfo if null from context.

emmiegit committed:
Handle link stats and null page info.

emmiegit committed:
Set default wikitext backend back to wikidot.

emmiegit committed:
Remove unused variable.

emmiegit committed:
Add missing import.

emmiegit committed:
Remove dead code.

emmiegit committed:
Remove commented out Outdater code.

emmiegit committed:
Make getWikitextBackend() a static method.

emmiegit committed:
Fix enum type.

emmiegit committed:
Fix issues with page saving.

emmiegit committed:
Rename 'null' -> 'dummy' in wikitext backend feature flag.

emmiegit committed:
Add missing use.

emmiegit committed:
Convert from native to FFI PageInfo.

emmiegit committed:
Fix DummyBackend.

emmiegit committed:
Fix misc issues.

Monkatraz reviewed (approved):
yikes

I'm so sorry

rossjrw commented on web/app/Services/Wikitext/HtmlMeta.php:
```diff
@@ -3,42 +3,24 @@
-
 /**
- * Class HtmlMeta, representing a returned 'struct ftml_html_meta' object.
+ * Class HtmlMeta, represents one &lt;meta&gt;, to be included in a rendered HTML document.
```
```suggestion
 * Class HtmlMeta, represents one <meta>, to be included in a rendered HTML document.
```

I assume PHP allows angle brackets in comments?

emmiegit replied:
No it interprets HTML elements as elements, like bolding a section of your phpdoc. Given that having a random open `<meta>` tag is not valid, we have to escape the angle brackets.

rossjrw commented on web/app/Services/Wikitext/HtmlOutput.php:
```diff
@@ -3,27 +3,45 @@
+     */
     public string $html;
+
+    /**
+     * @var string The CSS output of the rendering process.
+     *
+     * This concatenates separate CSS styles specified in the tree in one string.
+     * The styles are not necessarily valid or safe, as they come from whatever the user inputted.
+     */
     public string $style;
```
Out of scope for this PR, but noting that I think it'd be helpful to retain Wikidot's behaviour of putting each CSS snippet into separate style elements - created an issue [WJ-570](https://scuttle.atlassian.net/browse/WJ-570)

emmiegit replied:
I guess that's fair. I originally combined them to avoid extra string buffer allocations in Rust, but that's kind of moot given all the FFI and such happening.

emmiegit committed:
Fix HtmlOutput constructor.

emmiegit committed:
Fix name usages html -> body.

emmiegit committed:
Fix TextWikiBackend.

emmiegit committed:
Rename $html -> $body.

emmiegit committed:
Rename $linkStats -> $backlinks.

emmiegit commented:
thanks @pxdnbluesoul @rossjrw @Monkatraz
